src/Databases/DataLake/DatabaseDataLake.cpp

```diff
@@ -448,20 +448,39 @@ DatabaseTablesIteratorPtr DatabaseDataLake::getTablesIterator(
     const FilterByNameFunction & filter_by_table_name,
     bool /* skip_not_loaded */) const
 {
+    std::mutex mutex;
     Tables tables;
     auto catalog = getCatalog();
     const auto iceberg_tables = catalog->getTables();
 
+    auto & pool = Context::getGlobalContextInstance()->getIcebergCatalogThreadpool();
+
     for (const auto & table_name : iceberg_tables)
     {
         if (filter_by_table_name && !filter_by_table_name(table_name))
             continue;
 
-        auto storage = tryGetTable(table_name, context_);
-        [[maybe_unused]] bool inserted = tables.emplace(table_name, storage).second;
-        chassert(inserted);
+        try
+        {
+            pool.scheduleOrThrow(
+                [this, table_name, &tables, &mutex, context_]() mutable
+                {
+                    auto storage = tryGetTable(table_name, context_);
+                    std::lock_guard lock(mutex);
+                    [[maybe_unused]] bool inserted = tables.emplace(table_name, storage).second;
+                    chassert(inserted);
+                });
+        }
+        catch (...)
+        {
+            tryLogCurrentException(log, "Failed to schedule task");
+            pool.wait();
+            throw;
+        }
     }
 
+    pool.wait();
```

Comment:
This will wait all tasks from all queries that are currently using this pool, not just the current query.

Is it ok?

Reply:
Fixed